SwiftGit2/Repository.swift

```diff
@@ -558,4 +579,246 @@ final public class Repository {
 		let iterator = CommitIterator(repo: self, root: branch.oid.oid)
 		return iterator
 	}
+
+	// MARK: - Diffs
+
+	public func diff(for commit: Commit) -> Result<Diff, NSError> {
+		typealias Delta = Diff.Delta
+
+		guard !commit.parents.isEmpty else {
+			// Initial commit in a repository
+			return self.diff(from: nil, to: commit.oid)
+		}
+
+		var diffs = [OpaquePointer]()
+		var mergeDiff: OpaquePointer? = nil
+		for parent in commit.parents {
+			let diff: Result<OpaquePointer, NSError> = self.diff(from: parent.oid, to: commit.oid)
+			guard diff.error == nil else {
+				return Result<Diff, NSError>.failure(diff.error!)
+			}
+
+			diffs.append(diff.value!)
+
+			if mergeDiff == nil {
+				mergeDiff = diff.value!
+			} else {
+				let mergeResult = git_diff_merge(mergeDiff, diff.value)
+				guard mergeResult == GIT_OK.rawValue else {
+					return .failure(NSError(gitError: mergeResult, pointOfFailure: "git_diff_merge"))
+				}
+			}
+		}
+
+		for diff in diffs {
+			git_object_free(diff)
+		}
+
+		return .success(Diff(mergeDiff!))
+	}
+
+	/// Caller is responsible to free returned git_diff with git_object_free
+	private func diff(from oldCommitOid: OID?, to newCommitOid: OID?) -> Result<OpaquePointer, NSError> {
+		assert(oldCommitOid != nil || newCommitOid != nil, "It is an error to pass nil for both the oldOid and newOid")
+
+		var oldTree: OpaquePointer? = nil
+		if let oid = oldCommitOid {
+			let result = unsafeTreeForCommitId(oid)
+			guard result.error == nil else {
+				return Result.failure(result.error!)
+			}
+
+			oldTree = result.value
```

Comment:
Maybe follow this line with `defer { git_object_free(oldTree) }` (and same with `newTree` below).

The locality of that is nice, but it should also help us free memory if we return an error.

Reply:
:+1: